compiler/luci/service/src/Nodes/CircleBatchMatMul.cpp

```diff
@@ -30,4 +52,56 @@ luci::CircleNode *CloneNodeLet<CN::ABC>::visit(const luci::CircleBatchMatMul *no
   return cloned;
 }
 
+// BatchMatMulV2 supports broadcasting in the batch dimensions(BatchMatMul doesn't)
+// TODO Distinguish BatchMatMul and BatchMatMulV2
+loco::TensorShape sinf::Algorithm::visit(const luci::CircleBatchMatMul *node)
+{
+  const auto x = loco::must_cast<CircleNode *>(node->x());
+  const auto y = loco::must_cast<CircleNode *>(node->y());
+
+  const auto x_shape = sinf::circle_shape(x);
+  const auto y_shape = sinf::circle_shape(y);
+
+  uint32_t x_rank = x_shape.rank();
+  uint32_t y_rank = y_shape.rank();
+  assert(x_rank >= 2 && y_rank >= 2);
+
+  uint32_t max_rank = x_rank > y_rank ? x_rank : y_rank;
+  loco::TensorShape output_shape;
+  output_shape.rank(max_rank);
+
```

Comment:
If this kind of comment is helpful to understand, please let me know. 
I'll append it to the code. 

```c++
  // BatchMatMul shape inference rule works with two-part
  // 
  // 1) Batch dimensions part
  //    - Batch dimensions corresponds the to input_shape[:-2]
  //    - General broadcast rules are used to infer the shape of the output batch
  // 
  // 2) Contracting dimensions part
  //    - Contracting dimensions corresponds to the input_shape[-2:]
  //    - General matrix multiplication shape inference applied for this part, 
  //      which means '(x_lhs, x_rhs) x (y_lhs, y_rhs) => (x_lhs, y_rhs)'
```